extensions/s3/src/main/java/io/deephaven/extensions/s3/S3OutputStream.java

```diff
@@ -0,0 +1,268 @@
+    private final S3Instructions s3Instructions;
+
+    private final int partSize;
+    private final int numConcurrentParts; // TODO Better name for this
+
+    private final List<CompletedPart> completedParts;
+    private final List<OutgoingRequest> pendingRequests;
+
+    private int nextPartNumber;
+    private String uploadId;
+
+    S3OutputStream(
+            @NotNull final URI uri,
+            @NotNull final S3AsyncClient s3AsyncClient,
+            @NotNull final S3Instructions s3Instructions) {
+        this.uri = s3AsyncClient.utilities().parseUri(uri);
+        this.s3AsyncClient = s3AsyncClient;
+        this.s3Instructions = s3Instructions;
+
+        this.partSize = s3Instructions.partSize();
+        this.numConcurrentParts = s3Instructions.numConcurrentParts();
+        this.pendingRequests = new ArrayList<>(numConcurrentParts);
+
+        this.nextPartNumber = MIN_PART_NUMBER;
+        this.completedParts = new ArrayList<>();
+    }
+
+    public void write(int b) throws IOException {
+        write(new byte[] {(byte) b}, 0, 1);
+    }
+
+    public void write(byte[] b) throws IOException {
+        write(b, 0, b.length);
+    }
+
+    public void write(final byte @NotNull [] b, int off, int len) throws IOException {
+        while (len != 0) {
+            if (uploadId == null) {
+                // Initialize the upload ID for the multipart upload
+                uploadId = initiateMultipartUpload();
+            }
+
+            // We use request slots in a circular queue fashion
+            final int nextSlotId = (nextPartNumber - 1) % numConcurrentParts;
+            if (pendingRequests.size() == nextSlotId) {
+                pendingRequests.add(new OutgoingRequest(partSize));
+            } else if (pendingRequests.size() < nextSlotId) {
+                throw new IllegalStateException("Unexpected slot ID " + nextSlotId + " for uri " + uri + " with " +
+                        pendingRequests.size() + " pending requests.");
+            }
+
+            // Wait for the oldest upload to complete if no space is available
+            final OutgoingRequest useRequest = pendingRequests.get(nextSlotId);
+            if (useRequest.future != null) {
+                waitForCompletion(useRequest);
+            }
+
+            // Write as much as possible to this buffer
+            final ByteBuffer buffer = useRequest.buffer;
+            final int remaining = buffer.remaining();
+            if (remaining >= len) {
+                buffer.put(b, off, len);
+                if (!buffer.hasRemaining()) {
+                    sendPartRequest(useRequest);
+                }
+                break; // done
+            }
+            buffer.put(b, off, remaining);
+            sendPartRequest(useRequest);
+            off += remaining;
+            len -= remaining;
+        }
+    }
+
+    public void flush() throws IOException {
+        final int nextSlotId = (nextPartNumber - 1) % numConcurrentParts;
```

Comment:
We need to consider what to do if a user asks for a "short" flush. I think `flush()` should refuse to send a part smaller than the minimum, and we should have a `done()` method.

That is:
`done()` -> really flush, even if we're short. Blow up on subsequent writes.
`complete()` -> complete the upload
`rollback()` -> no-op for s3